### PR TITLE
Add types for contrib.stochastic_support and improve the HSGP module

### PR DESCRIPTION
**Pull request by juanitorduz**
Related to https://github.com/pyro-ppl/numpyro/issues/299

### CODE REVIEW

File: numpyro/contrib/stochastic_support/dcc.py [hunk 1]
```diff
@@ -3,6 +3,9 @@
 
 from abc import ABC, abstractmethod
 from collections import OrderedDict, namedtuple
+from typing import Any, Callable, OrderedDict as OrderedDictType
+
+from jaxlib.xla_extension import ArrayImpl
```
Comment (tillahoffmann) [inline numpyro/contrib/stochastic_support/dcc.py:8]: Regarding the default typing, what are the implications of using `ArrayImpl` vs `jax.numpy.ndarray`, for example?

Comment (juanitorduz) [reply numpyro/contrib/stochastic_support/dcc.py:8]: This is a great question! Thanks for the feedback! So we started trying with type-hints with `ArrayImpl` in https://github.com/pyro-ppl/numpyro/pull/1819#discussion_r1649747153 

On the other hand in https://jax.readthedocs.io/en/latest/jax.typing.html its suggested to use `jax.typing.ArrayLike`. However when I use this for `x: ArrayLike` and have inside the function `x.ndim`, `mypy` complains with 

```python
error: Item "builtins.bool" of "Array | ndarray[Any, Any] | numpy.bool | number[Any] | builtins.bool | int | float | complex" has no attribute "ndim"  [union-attr]
```

Hence, this is why I am continuing with `ArrayImpl` and waiting for feedback :)

Comment (fehiepsi) [reply numpyro/contrib/stochastic_support/dcc.py:8]: I would like to avoid importing stuff from jaxlib. I guess it is better to use `jax.Array`

Comment (fehiepsi) [reply numpyro/contrib/stochastic_support/dcc.py:8]: Re ndim: you can use jnp.ndim(x), which will accept ArrayLike instances, including python scalars.

Comment (juanitorduz) [reply numpyro/contrib/stochastic_support/dcc.py:8]: Great input! I'll try that then 🙂!

**Comment by juanitorduz** [comment]
ok! After changing the types (and also updating https://github.com/pyro-ppl/numpyro/pull/1906). I think this is the way to go as suggested ( I even caught minor issues) :) 

As always, feedback is welcome!

**Review by fehiepsi — APPROVED** [review]
Looks great! I like ArrayLike -> Array, looks consistent with jax api. If not much work, let's also replace jnp.ndarray by jax.Array (optional).

**Pull request merged**